README.md
# KeypointNet

**Full dataset is available now!** KeypointNet is a large-scale and diverse 3D keypoint dataset that contains
83,231 keypoints and 8,329 3D models from 16 object categories, by leveraging numerous human annotations, based on ShapeNet models. Our paper is available on https://arxiv.org/pdf/2002.12687.pdf and is accepted to CVPR 2020. **You can explore our dataset on <a href="https://qq456cvb.github.io/keypointnet/explore/" target="_blank">https://qq456cvb.github.io/keypointnet/explore/</a>**.

# Change Logs
For updated dataset information, see [Change Log](CHANGELOG.md)

# Keypoint Data
Dataset can be downloaded from <a href="https://drive.google.com/drive/folders/1_d1TzZEF25Wy5kRj5ZugrgGeyf7xxu8F?usp=sharing" target="_blank">Google Drive</a>  or <a href="https://1drv.ms/u/s!Aj0NuSsDz6hDyF3LT3xaPkXK9DXC?e=kcrfSg" target="_blank">OneDrive</a>. Annotated JSON data is placed under **annotations**. In addition, we provide sampled point clouds (2048 points) for each ShapeNet model under **pcds**.

We have processed and cleaned labels for airplane (1022 models), bathtub (492 models), bed (146 models), bottle (380 models), cap (38 models), car (1002 models), chair (999 models), guitar (697 models), helmet (90 models), knife (270 models), laptop (439 models), motorcycle (298 models), mug (198 models), skateboard (141 models), table (1124 models) and vessel (910 models).

**UPDATE:** we have managed to add **color** information onto sampled point clouds and keypoints. In addition, since processing raw ShapeNet obj file as colored triangle meshes is painful, we have generated corresponding ply files (named **ShapeNetCore.v2.ply.zip**) with vertex colors (diffuse texture color), for those are interested in dealing with triangle meshes. We believe color is an important source when learning from 3D geometries.

<img src="examples/captures/pcd.png" width="220" height="360" /><img src="examples/captures/obj.png" width="220" height="360" /><img src="examples/captures/ply.png" width="220" height="360" />
<!-- ![pcd](examples/captures/pcd.png){:height="360px" width="160px"}
![obj](examples/captures/obj.png){:height="360px" width="160px"}
![ply](examples/captures/ply.png){:height="360px" width="160px"} -->

## Data format
```javascript
[
    ...,
    {  
        "class_id": "03001627",  // WordNet id
        "model_id": "88382b877be91b2a572f8e1c1caad99e",  // model id
        "keypoints": [
            {
                "xyz": [0.16, 0.1, 0.1],  // xyz coordinate of keypoint
                "rgb": [255, 255, 255], // rgb color of keypoint, uint8
                "semantic_id": 0,  // id of semantic meaning
                "pcd_info": {
                    "point_index": 0  // keypoint index on corresponding point cloud
                },
                "mesh_info": {  // mesh information for both obj and ply files
                    "face_index": 0,  // index of mesh face where keypoint lies
                    "face_uv": [0.2, 0.4, 0.4]  // barycentric coordinate on corresponding mesh face
                }
            },
            ...
        ],
        "symmetries": { // information of keypoint symmetries
            "reflection": 
            [
                {
                    "kp_indexes": [0, 1]  // keypoint indexes of a reflection symmetric group
                },
                ...
            ],
            "rotation":
            [
                {
                    "kp_indexes": [0, 1, 2, 3],  // keypoint indexes of a rotation symmetric group
                    "is_circle": true,  // true if this rotation symmtric group is a rounding circle
                    "circle": {
                        "center": [0.2, 0.5, 0.2],  // circle center
                        "radius": 0.32,  // circle radius
                        "normal": [0, 1.0, 0],  // normal of circle plane
                    }
                },
                ...
            ]
        }

    },
    ...
]
```

# Examples
Example scripts on reading and visualizing keypoints on both point clouds and triangle meshes are placed under **examples**.

# Keypoint Detection Tasks
Keypoint saliency and correspondence training and evaluation baselines for various backbones are placed under **benchmark_scripts**. For more details, please refer to **benchmark_scripts/README.md**.

# Data Splits
train/val/test splits are placed under **splits**. Each line is formatted as `[class_id]-[model_id]`.


# Citation
If you find KeypointNet data or code useful in your research, please consider citing:
```
@article{you2020keypointnet,
  title={KeypointNet: A Large-scale 3D Keypoint Dataset Aggregated from Numerous Human Annotations},
  author={You, Yang and Lou, Yujing and Li, Chengkun and Cheng, Zhoujun and Li, Liangwei and Ma, Lizhuang and Lu, Cewu and Wang, Weiming},
  journal={arXiv preprint arXiv:2002.12687},
  year={2020}
}
```

# TODOs

- [x] clean labels for more classes
- [x] add colored pcds/meshes
- [x] a browser interface to explore dataset


CHANGELOG.md
# Change Log

## 2021-01-05
- Fixed problematic annotations for three chairs. Removed reflection planes of legs in swivel chairs.
## 2020-12-12
- Done code cleanup. Now it's much easier to run and read benchmark code.

## 2020-06-15
- **Full dataset is available now!** Note: some of knives are misaligned (where x-axis is flipped) in the orginal ShapeNet. We picked those misaligned ones and stored them into **knife_misaligned.txt**.

## 2020-04-18
- We've split the semantic ids of normal chairs (**four** legs) and swivel chairs (**five** legs with rotational symmetries). **Before**, the four legs of normal chairs have semantic id: **10, 11, 12, 13** and the five legs of swivel chairs have semantic id: **10, 11, 12, 13, 14**. **Now**, the four legs of normal chairs have semantic id: **17, 18, 19, 20** and semantic ids of swivel chairs are the same as before.

## 2020-04-17
- cleaned saliency experiments code
- cleaned correspondence experiments code

## 2020-04-03

- Add colored pcd and ply file on ShapeNet.v2
- Replace json annotations with updated information

## 2020-04-01

- First release on ShapeNet.v2 airplane, chair and table classes


benchmark_scripts/dataset.py
import numpy as np
from glob import glob
import os
import torch
import json
from scipy.stats import special_ortho_group
from sklearn.metrics import pairwise_distances_argmin


BASEDIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..')
ID2NAMES = {"02691156": "airplane",
            "02808440": "bathtub",
            "02818832": "bed",
            "02876657": "bottle",
            "02954340": "cap",
            "02958343": "car",
            "03001627": "chair",
            "03467517": "guitar",
            "03513137": "helmet",
            "03624134": "knife",
            "03642806": "laptop",
            "03790512": "motorcycle",
            "03797390": "mug",
            "04225987": "skateboard",
            "04379243": "table",
            "04530566": "vessel",}

NAMES2ID = {v: k for k, v in ID2NAMES.items()}


def naive_read_pcd(path):
    lines = open(path, 'r').readlines()
    idx = -1
    for i, line in enumerate(lines):
        if line.startswith('DATA ascii'):
            idx = i + 1
            break
    lines = lines[idx:]
    lines = [line.rstrip().split(' ') for line in lines]
    data = np.asarray(lines)
    pc = np.array(data[:, :3], dtype=np.float)
    colors = np.array(data[:, -1], dtype=np.int)
    colors = np.stack([(colors >> 16) & 255, (colors >> 8) & 255, colors & 255], -1)
    return pc, colors


def add_noise(x, sigma=0.015, clip=0.05):
    noise = np.clip(sigma*np.random.randn(*x.shape), -1*clip, clip)
    return x + noise


def normalize_pc(pc):
    pc = pc - pc.mean(0)
    pc /= np.max(np.linalg.norm(pc, axis=-1))
    return pc


class KeypointSaliencyDataset(torch.utils.data.Dataset):
    def __init__(self, cfg, split):
        super().__init__()
        self.catg = cfg.class_name
        self.cfg = cfg
        
        annots = json.load(open(os.path.join(BASEDIR, cfg.data.annot_path)))
        annots = [annot for annot in annots if annot['class_id'] == NAMES2ID[cfg.class_name]]
        keypoints = dict([(annot['model_id'], [kp_info['pcd_info']['point_index'] for kp_info in annot['keypoints']]) for annot in annots])
        
        split_models = open(os.path.join(BASEDIR, "splits/{}.txt".format(split))).readlines()
        split_models = [m.split('-')[-1].rstrip('\n') for m in split_models]
        
        self.pcds = []
        self.keypoints = []
        self.mesh_names = []
        for fn in glob(os.path.join(BASEDIR, cfg.data.pcd_root, NAMES2ID[cfg.class_name], '*.pcd')):
            model_id = os.path.basename(fn).split('.')[0]
            if model_id not in split_models:
                continue
            self.keypoints.append(keypoints[model_id])
            self.pcds.append(naive_read_pcd(fn)[0])
            self.mesh_names.append(model_id)

        self.nclasses = 2
        self.split = split
    
    def __getitem__(self, idx):
        pc = self.pcds[idx]
        mesh_name = self.mesh_names[idx]
        label = self.keypoints[idx]
        bin_label = np.zeros((pc.shape[0],), dtype=np.int64)
        bin_label[label] = 1
        
        if self.cfg.normalize_pc:
            pc = normalize_pc(pc)
        
        if self.split == 'train':
            if self.cfg.augmentation.gaussian_noise:
                pc = add_noise(pc, sigma=0.004, clip=0.01)

            if self.cfg.augmentation.translation:
                tr = np.random.rand() * 0.02
                rot = special_ortho_group.rvs(3)
                pc = pc @ rot
                pc += np.array([[tr, 0, 0]])
                pc = pc @ rot.T

            # allow rotation along gravity axis
            if self.cfg.augmentation.rot_gravity:
                angle = np.random.rand() * 2 * np.pi
                rot = np.array([[np.cos(angle), 0, -np.sin(angle)], [0, 1, 0], [np.sin(angle), 0, np.cos(angle)]])
                pc = (rot @ pc.T).T
                
        return pc.astype(np.float32), bin_label, mesh_name

    def __len__(self):
        return len(self.pcds)
    

class KeypointCorrespondenceDataset(torch.utils.data.Dataset):
    def __init__(self, cfg, split):
        super().__init__()
        self.catg = cfg.class_name
        self.cfg = cfg
            
        annots = json.load(open(os.path.join(BASEDIR, cfg.data.annot_path)))
        annots = [annot for annot in annots if annot['class_id'] == NAMES2ID[cfg.class_name]]
        keypoints = dict([(annot['model_id'], [(kp_info['pcd_info']['point_index'], kp_info['semantic_id']) for kp_info in annot['keypoints']]) for annot in annots])
        rotation_groups = dict([(annot['model_id'], annot['symmetries']['rotation']) for annot in annots])
        
        self.nclasses = max([max([kp_info['semantic_id'] for kp_info in annot['keypoints']]) for annot in annots]) + 1
        self.split = split
        
        split_models = open(os.path.join(BASEDIR, "splits/{}.txt".format(split))).readlines()
        split_models = [m.split('-')[-1].rstrip('\n') for m in split_models]
        
        self.pcds = []
        self.keypoints = []
        self.mesh_names = []
        self.rotation_infos = []
        self.idx2semids = []
        for fn in glob(os.path.join(BASEDIR, cfg.data.pcd_root, NAMES2ID[cfg.class_name], '*.pcd')):
            model_id = os.path.basename(fn).split('.')[0]
            if model_id not in split_models:
                continue
            
            idx2semid = dict()
            curr_keypoints = -np.ones((self.nclasses,), dtype=np.int)
            for i, kp in enumerate(keypoints[model_id]):
                curr_keypoints[kp[1]] = kp[0]
                idx2semid[i] = kp[1]
            self.keypoints.append(curr_keypoints)
            self.rotation_infos.append(rotation_groups[model_id])
            self.idx2semids.append(idx2semid)
            self.pcds.append(naive_read_pcd(fn)[0])
            self.mesh_names.append(model_id)

    def __getitem__(self, idx):
        pc = self.pcds[idx]
        label = self.keypoints[idx]
        keypoint_coords = np.array([pc[idx] for idx in label if idx >= 0])
        rotation_info = self.rotation_infos[idx]
        idx2semid = self.idx2semids[idx]
        mesh_name = self.mesh_names[idx]
        
        if self.cfg.normalize_pc:
            pc = normalize_pc(pc)
        
        labels = [label]
        if len(rotation_info) > 0 and not self.catg == 'chair':  # rotational axis exists
            group_idxs = [info['kp_indexes'] for info in rotation_info]
            all_circle = not (False in [info['is_circle'] for info in rotation_info])
            
            group = group_idxs[0]
            group_points = np.array([pc[label[idx2semid[i]]] for i in group])
            rotation_center = np.mean(group_points, 0)
            rotation_center[1] = 0  # set y to 0
            rotation_multiplex = len(group)
            if all_circle:
                rotation_multiplex = 36
                
            for i in range(1, rotation_multiplex):
                angle = 2 * np.pi / rotation_multiplex * i
                keypoints_rotated = (keypoint_coords - rotation_center) @ np.array([[np.cos(angle), 0, -np.sin(angle)], [0, 1, 0], [np.sin(angle), 0, np.cos(angle)]]).T + rotation_center
                label_rotated = -np.ones_like(label, dtype=np.int)
                label_rotated[label >= 0] = pairwise_distances_argmin(keypoints_rotated, pc)
                
                labels.append(label_rotated)
        
        if self.split == 'train':
            if self.cfg.augmentation.gaussian_noise:
                pc = add_noise(pc, sigma=0.004, clip=0.01)

            if self.cfg.augmentation.translation:
                tr = np.random.rand() * 0.02
                rot = special_ortho_group.rvs(3)
                pc = pc @ rot
                pc += np.array([[tr, 0, 0]])
                pc = pc @ rot.T

            # allow rotation along gravity axis
            if self.cfg.augmentation.rot_gravity:
                angle = np.random.rand() * 2 * np.pi
                rot = np.array([[np.cos(angle), 0, -np.sin(angle)], [0, 1, 0], [np.sin(angle), 0, np.cos(angle)]])
                pc = (rot @ pc.T).T
            
        return pc.astype(np.float32), labels, mesh_name

    def __len__(self):
        return len(self.pcds)
